DataStructures/BinarySearchTree/BinarySearchTree.cs

```diff
@@ -0,0 +1,421 @@
+        /// Insert a key into the BST.
+        /// </summary>
+        /// <param name="key">The key to insert.</param>
+        /// <exception cref="ArgumentNullException">
+        /// Thrown if key is null.
+        /// </exception>
+        /// <exception cref="ArgumentException">
+        /// Thrown if key is already in BST.
+        /// </exception>
+        public void Add(TKey key)
+        {
+            if (key == null)
+            {
+                throw new ArgumentNullException("Key cannot be null!");
+            }
+
+            if (root == null)
+            {
+                root = new BinarySearchTreeNode<TKey>(key);
+            }
+            else
+            {
+                Add(root, key);
+            }
+
+            Count++;
+        }
+
+        /// <summary>
+        /// Insert multiple keys into the BST.
+        /// Keys are inserted in the order they appear in the sequence.
+        /// </summary>
+        /// <param name="keys">Sequence of keys to insert.</param>
+        /// <exception cref="ArgumentException">
+        /// Thrown if a key in the sequence is already in BST.
+        /// </exception>
+        public void AddRange(IEnumerable<TKey> keys)
+        {
+            foreach (TKey key in keys)
+            {
+                Add(key);
+            }
+        }
+
+        /// <summary>
+        /// Find a node with the specified key.
+        /// </summary>
+        /// <param name="key">The key to search for.</param>
+        /// <returns>The node with the specified key if it exists, otherwise a default value is returned.</returns>
+        /// <exception cref="ArgumentNullException">
+        /// Thrown if key is null.
+        /// </exception>
+        public BinarySearchTreeNode<TKey>? Search(TKey key)
+        {
+            if (key == null)
+            {
+                throw new ArgumentNullException("Key cannot be null!");
+            }
+
+            return Search(root, key);
+        }
+
+        /// <summary>
+        /// Checks if the specified key is in the BST.
+        /// </summary>
+        /// <param name="key">The key to search for.</param>
+        /// <returns>true if the key is in the BST, false otherwise.</returns>
+        /// <exception cref="ArgumentNullException">
+        /// Thrown if key is null.
+        /// </exception>
+        public bool Contains(TKey key)
+        {
+            if (key == null)
+            {
+                throw new ArgumentNullException("Key cannot be null!");
+            }
+
+            return Search(root, key) != null;
+        }
+
+        /// <summary>
+        /// Removes a node with the specified key from the BST.
+        /// </summary>
+        /// <param name="key">The key to remove.</param>
+        /// <returns>true if the operation was successful, false otherwise.</returns>
+        /// <exception cref="ArgumentNullException">
+        /// Thrown if key is null.
+        /// </exception>
+        public bool Remove(TKey key)
```

Comment:
Codacy found an issue: [The Cyclomatic Complexity of this method is 15 which is greater than 10 authorized.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6277689)